### DISCUSSION TITLE
[RFC] StaticTable for fast perf & virtual scroll support

### DISCUSSION THREAD

**Discussion by zombieJ** [ant-design/ant-design#41500] [RFCs - archive]
# 📋 RFC

## Summary

Table provide `Table.StaticTable` for no additional render & default virtual scroll support.

## Basic example

```tsx
import { Table } from `antd`;

const columns = [
  {
    dataIndex: 'name',
    fixed: 'left',
    width: 100, // Must have since always default using virtual scroll
  },
];

render(<Table.StaticTable dataSource={...} columns={columns} />);
```

## Motivation

ref #20339

Table requires large data usage in some case. It's hard for developer to optimize since cell need full render. It can be use `shouldCellUpdate` for cell memo but can not avoid first time full render.

## Detailed

Follow all the props current have but require `width` for every column. Using virtual grid and fixed ability mixing for fixed column support.

|  | Table | StaticTable |  |
| --- | --- | --- | --- |
| Virtual scroll | ⛔️ | ✅ | Table need customize `components.body` to support this |
| fixed column |  ✅ |  ✅ | |
| Dynamic `column.render` | ✅ | ⛔️ | StaticTable will not re-render cell when `dataSource` & `column` not changed  |
| Cell autoSize | ✅ | ⛔️ | StaticTable using virtual scroll which means it will not fully render all the content in view. So `column.width` is must have |
| Pagination | ✅ | ⛔️ | StaticTable using virtual scroll which is always fully render |
| components | ✅ | ⛔️ | Virtual Table cell is not real TableCell |
| Form support | ✅ | ⛔️ | StaticTable do not render all the cell which means Form field is also not exist in the view |
| Summary | ✅ | ✅ |  |
| Expandable | ✅ | ✅ | |
| Checkable |✅ |  ✅ |  |

## Drawbacks

Not support auto width

## Adoption strategy

Extends `rc-table` to provide Static ability.

**Comment by Alainterieur85** [comment]
Nice

**Comment by crazyair** [comment]
奈斯

**Comment by afc163** [comment]
需要有个表格对 StaticTable 和 Table 进行对比，说清楚 pros and cons

**Comment by AsuraKev** [comment]
@zombieJ Cant we do anything about the Table component or must we opt into the Static Table when we need performance? Will we lose some Table features if we use Static table? thanks

**Comment by crazyair** [comment]
`Pagination ` 有可能每页200条，也需要分页

**Comment by AsuraKev** [comment]
what if you virtual scroll and each scroll is 200+ record, isnt that gonna still be the same issue with performance?

**Reply by crazyair** [reply]
What I mean is, need to support pagination

**Comment by AsuraKev** [comment]
Is this RFC going to be the resolution for slow performance of row selection of large dataset? Whats gonna be left for the main Table component

**Comment by dreambold** [comment]
So does it require loading all the data at a time as in the official doc? https://ant.design/components/table#components-table-demo-virtual-list

It's not practical and you'd better add some features to convert the pagination table into infinite scroll easily.
Most of the current plugins/packages work this way loading data pagewise on scroll. Something like this: https://www.npmjs.com/package/virtualizedtableforantd4

I was going to use this virtual table, but I found it's not practical eventually unless it has features to adapt to the pagination tables. Most of the APIs are built that way, page-wise data.
